x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ParseTables.java

```diff
@@ -0,0 +1,244 @@
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.breaker.NoopCircuitBreaker;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.BytesRefBlock;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.LongBlock;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.xcontent.XContentParseException;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.esql.Column;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Parses the {@code tables} request body parameter.
+ */
+public class ParseTables {
+    public static final Set<DataType> SUPPORTED_TYPES = Set.of(DataTypes.INTEGER, DataTypes.KEYWORD, DataTypes.LONG);
+    private static final int MAX_LENGTH = (int) ByteSizeValue.ofMb(1).getBytes();
+
+    private final BlockFactory blockFactory;
+    private final EsqlQueryRequest request;
+    private final XContentParser p;
+    private int length;
+
+    ParseTables(EsqlQueryRequest request, XContentParser p) {
+        // TODO use a real block factory
+        // NOCOMMIT really account for size
+        this.blockFactory = new BlockFactory(new NoopCircuitBreaker(CircuitBreaker.REQUEST), BigArrays.NON_RECYCLING_INSTANCE);
+        this.request = request;
+        this.p = p;
+    }
+
+    void parseTables() throws IOException {
+        if (p.currentToken() != XContentParser.Token.START_OBJECT) {
+            throw new XContentParseException(p.getTokenLocation(), "expected " + XContentParser.Token.START_OBJECT);
+        }
+        while (true) {
+            switch (p.nextToken()) {
+                case END_OBJECT -> {
+                    return;
+                }
+                case FIELD_NAME -> {
+                    String name = p.currentName();
+                    p.nextToken();
+                    request.addTable(name, parseTable());
+                }
+            }
+        }
+    }
+
+    private Map<String, Column> parseTable() throws IOException {
+        Map<String, Column> columns = new TreeMap<>();
+        boolean success = false;
+        try {
+            if (p.currentToken() != XContentParser.Token.START_OBJECT) {
+                throw new XContentParseException(p.getTokenLocation(), "expected " + XContentParser.Token.START_OBJECT);
+            }
+            while (true) {
+                switch (p.nextToken()) {
+                    case END_OBJECT -> {
+                        success = true;
+                        return columns;
+                    }
+                    case FIELD_NAME -> {
+                        // NOCOMMIT just infer type from the json
```

Comment:
I'm not entirely sure we want this. I mean, maybe. But stuff like `DATE`s would be nice to be able to parse one time.

Reply:
This open the discussion about what is the data format allowed on the wire (is CSV, TSV, arrow, etc...)?
IF we end up going this route, I expect we'll have a qualifier to indicate the resource type (CSV, etc..).